### PR TITLE
RENO-3008: Upgrade to React 17 + NextJS 12

### PR DESCRIPTION
**Pull request by nypl-wluisi**
[Jira Ticket](https://jira.nypl.org/browse/RENO-3008)

**This PR does the following:**
- Updates to React 17
- Updates to NextJS 12.2.0
- Various updates to support React 17 and NextJS 12.

### Review Steps:
- [ ] Create a new project in VS Code for "NextJS 12" -- There are large changes here, so it'll be easier to have this as a separate instance locally.
- [ ] Pull in code
- [ ] Run `npm install`
- [ ] Run `npm run build && npm start`
- [ ] Run unit tests `npm test`
- [ ] Run cypress tests `npm run cypress`

### Verify

Verify that there are no visual or functional regressions on the following pages:
- http://localhost:3000/locations
- http://localhost:3000/locations/request-visit
- http://localhost:3000/blog (and all related pages)
- http://localhost:3000/research/collections/articles-databases (and all related pages)

### CODE REVIEW

**Comment by nypl-wluisi** [comment]
@isastettler Hey, this is ready for final review. I made a few more changes today to fix some of the failing tests. See updated instructions in the description (mainly 2 more steps to run both sets of tests locally, both should pass 100%)

File: src/pages/research/collections/articles-databases/featured/[slug].tsx [hunk 1]
```diff
@@ -1,25 +1,33 @@
-import React from "react";
+import * as React from "react";
```
Comment (isastettler) [inline src/pages/research/collections/articles-databases/featured/[slug].tsx:1]: Is this your preferred way of importing React? I noticed you changed it in a few files and used React.useEffect in file ...articles-databases/[slug].tsx

File: src/pages/blog/[...slug].tsx [hunk 2]
```diff
@@ -1,17 +1,22 @@
-import React from "react";
+import * as React from "react";
 // Next
+import { GetServerSidePropsContext } from "next";
 import { useRouter } from "next/router";
 import Error from "./../_error";
 // Apollo
 import { gql, useQuery } from "@apollo/client";
-import { withApollo } from "../../apollo/client/withApollo";
+import withApollo from "./../../apollo/withApollo";
```
Comment (isastettler) [inline src/pages/blog/[...slug].tsx:8]: What is the difference between the withApollo.js file in the client folder and the withApollo.tsx file in withApollo?

**Review by isastettler — APPROVED** [review]
Looks all good! I'd love to go over the comments with you if time allows to get a better understanding for myself.

**Pull request merged**